Feature/typescript integrated typings

alexandercerutti opened the pull request:
Hello there, I'm working on a migration to Typescript of a library that uses a very old version of `bl` (`v1.2.3` - which, for unknown reasons, is unlisted on NPM to me, but still accessible via URL).

I'm not a very huge fan of DefinitelyTyped but I _DefinitelyLove_ Typescript. Also, I was looking at `@types/bl`, and they seemed a bit wrong to me, based on the implementation.

So I decided to rewrite them with a better structure to represent how the code was written (in fact `interfaces` allow to specify of both a constructor and a function to be executed) and inside the package.

Now all the methods include a documentation comment based on the one in the Readme.

Hope this is appreciated. I also tried to run the tests available inside `@types/bl`, and they seem to be okay, and run linting.

I hope this is appreciated. Let me know if there's anything that you feel is wrong, and I'll _feeex_ it! 🛠️

mcollina commented:
wdyt @rvagg? Should we ship types?

rvagg commented:
Yeah, I don't mind shipping types, but checking correctness of manually generated types and keeping them in sync is the challenge. I've mostly been going with the approach of annotations and using tsc to compile type definitions from my JS and ship that. But doing that means making sure the annotations are complete and correct through the whole codebase, which can be a bit tedious.

We could ship this and cross fingers, and hope that people show up in the future to adjust and correct the over time, but it'll always be a catch-up effort as TypeScript evolves and our dependencies evolve too.

mcollina commented:
@alexandercerutti could you add some https://www.npmjs.com/package/tsd tests?

alexandercerutti commented:
@mcollina @rvagg I've chosen a different approach, which I think is more reliable in terms of testing types: type-checking existing tests.

Luckily for us, Typescript supports a subset of JSDoc, so I sightly changed `test/test.js` to support type-checking, added typescript as a dev dependency, and a new script in `package.json`, to check the file without emitting any compiled file but just errors, in case.

By doing so, I also discovered some cases that were not supported by the declarations I created.

For example, methods like `BufferList.readIntLE` were not supporting the second parameter (and they still don't, based on `@types/bl`), but they should, so I changed their signature to inherit it from `NodeJS.Buffer`.

I also added types support for BufferList and BufferListStream to `Uint8Array` and `number`.

Just a few notes and questions:

- Since tests use a "private" variable `BufferListStream._bufs` to check the correctness, I had to create a `@typedef` to add it to `BufferListStream` without exposing it inside typescript typings. This might become a little critical point in case of changes IMHO. I don't really think that is very correct to use a private variable in tests, but I cannot come out with a different solution that is not a getter.

- I noticed that, for what concerns `duplicate` test, `bl.prototype` and `dup.prototype` is checked.

https://github.com/rvagg/bl/blob/ccff4dc53facd29bca22b6f60031986c9c7c87fd/test/test.js#L794-L797

But this test is always to be `true` because both are undefined. In fact, BufferStreamList merges the prototype of BufferList, but it not extends (was here the concern that `Object.setPrototypeOf()` [is "slow"](https://stackoverflow.com/questions/52654781/why-does-using-object-setprototypeof-impact-performance-heavily) ?).
So to add a type for the `prototype`, I set a generic "Object" for both `BufferList` and `BufferListStream`.

- Just for a matter of coherence, I renamed `BufferList` in `BufferListStream` import in `test/test.js`, cause the default one is `BufferListStream`. It was confusing.

- I noticed that a `package-lock.json` is missing here. Should it get added (maybe in a different PR)?

- I noticed that this repo misses a configuration setting (like prettier or `.editorconfig`). I had to challenge my editor to attempt to keep styles as coherent as possible (it was changing everything in tests and typings). Should a configuration get added (maybe in a different PR)?

Let me know what you think about this, any concerns, or if I should revert this at all and proceed with `tsd` tests.

rvagg commented on package.json:
```diff
@@ -6,7 +6,8 @@
     "lint": "standard *.js test/*.js",
-    "test": "npm run lint && node test/test.js | faucet"
+    "test": "npm run lint && npm run test:typings && node test/test.js | faucet",
+    "test:typings": "npx tsc --allowJs --noEmit test/test.js"
```
```suggestion
    "test:types": "npx tsc --allowJs --noEmit test/test.js"
```
really minor nit, just feels more consistent across repos I personally work with (maybe there's a developing convention for the word "typings" that I've not seen though?)

alexandercerutti replied:
It is okay to me to change it! I'm not going to accept suggestion (as Github suggests) just because I need to update also the `test` script to use `test:types` instead of `test:typings`, so I'll commit it soon 👍

rvagg commented:
some brief drive-by thoughts while I have some time to glance at this:

* no `package-lock.json` please
* no `.editorconfig` please
* I _think_ the `Object.assign()` looks right? it's copying properties from one prototype to the other, not making an inheritance or equivalence
* the duplicate test does look wrong, I think the assertion should probably be `t.same(Object.getPrototypeOf(bl), Object.getPrototypeOf(dup))`
* I think you should just be able to `// @ts-ignore` the `_bufs` usage in the test to make it not be concerned about that, I do this pretty regularly when I'm running TS across JS tests that are doing funky things that you normally wouldn't do in TypeScript - it's pretty common to be testing edges of these kinds of things in JS that you wouldn't do in TS because you have strong asssumptions (like .. in my JS I tend to build in type checking `typeof foo === 'string'` kind of thing, but TS folks don't tend to do that because `foo:string`, but when you use that from JS you lose all guarantees, you really should be checking, so in tests you should be throwing numbers and objects to make sure it handles wrong types properly).
* `faucet` is getting in the way of test output for you, see the `"test"` script that pipes through it. Just run `node test/test.js` to get raw TAP output, and `tape.only()` the test to run just that one.

Otherwise I think this seems reasonable? I still wouldn't mind doing a compilation of the types from annotated source, I do this on most of my new projects these days and have been slowly retrofitting older ones I use a lot. https://github.com/rvagg/iamap is an example of this - `npm run build:types` will regenerate the `types/` directory. There's a basic `interface.ts` in there that has some things that are best described outside of annotations but otherwise the definitions are all pulled from the JS source annotations.

I'm not saying this is necessary here, it could be a stretch goal, but we may find the output differs when we do it like that.

alexandercerutti commented:
@rvagg thanks for the reply

- Just to ask you: why no `package-lock.json` and something like a `.editorconfig`?
- Okay yeah. I actually never thought that copying could be a solution actually. I always used `Object.create` in case.
- After a test, both `bl` and `dup` result in `Duplex` objects with methods. Yep, that feels correct. I'll open a new PR after this.
- I actually don't really like using `@ts-ignore` if I can avoid it. I mean, I've used it for an issue with `BufferEncoding` in node types because that's out of my control, but I'd avoid using it because it might cause problems in the future (if errors get suppressed, types checks might go error and we could not know that - also that would mean to suppress the whole line, so there could happen other errors).

> in my JS I tend to build in type checking typeof foo === 'string' kind of thing, but TS folks don't tend to do that because foo:string

This is a very bad behavior from ts folks - assuming that Typescript can completely replace JS - I agree with you.
That said, I think that typing just through JSDoc is a good compromise that does not expose `_bufs` and that allows us to be sure everything works as expected.

- I usually like to generate declarations through `tsc` too. I tried to make them be generated, but as I expected TS cannot infer correctly that `BufferListStream` "extend" `Duplex` and `BufferList`. I've looked a bit at your `iamap` but it seems to not have such cases, so your technique is valid as well.

That would be possible in this library, maybe, by using classes (but that would mean to sacrifice `BufferList()` and `BufferListStream()` I guess) and by not using `inherits` anymore ([which is considered legacy now](https://nodejs.org/api/util.html#utilinheritsconstructor-superconstructor)).

mcollina reviewed (approved):
lgtm

alexandercerutti commented:
Hey @rvagg, just to remember you about this :)

mcollina commented:
@rvagg I can release if you are ok with the approach.

piranna commented:
Any update on this? What's missing to be merged?

rvagg commented on package.json:
```diff
@@ -6,7 +6,8 @@
     "lint": "standard *.js test/*.js",
-    "test": "npm run lint && node test/test.js | faucet"
+    "test": "npm run lint && npm run test:types && node test/test.js | faucet",
+    "test:types": "npx tsc --allowJs --noEmit test/test.js"
```
```suggestion
    "test:types": "tsc --allowJs --noEmit test/test.js"
```

since we have `typescript` as a dependency

alexandercerutti replied:
Well, okay. I mean, I'm used to always add `npx` here just to give priority on `node_modules` over possible local installation. Sometimes this happened to me, probably while using `nvm`, but okay 👍

rvagg replied:
2 things - (1) this is the only good reason to install `typescript` as a devDep, if it's not being used in a script I don't believe we have another use for it, and (2) to my point about typescript breaking too much, `npx` is typically going to pull in the latest (unless you have a global installed) and this can break really easily and disrupt the normal workflow for others who then have to drop into typescript-land (where they may not be used to dealing with) to fix things unrelated to their changes.. I've seen this a bit too often with typescript on JS projects that I'm very weary about versioning.

rvagg commented on package.json:
```diff
@@ -30,8 +31,10 @@
     "standard": "^16.0.3",
-    "tape": "^5.2.2"
+    "tape": "^5.2.2",
+    "typescript": "^4.7.3"
```
```suggestion
    "typescript": "~4.7.3"
```

typescript breaks too much between semver-minor releases, especially when you're just consuming it to check JS

rvagg commented:
sorry, seems fine to me I think, just some minor things in package.json and we can get this out

alexandercerutti commented:
@rvagg done!
What about if, after this PR, we do set up a prettier or a `.editorconfig`? I think I mixed up a little bit the syntax due to my default config in vscode. 🤔

rvagg commented:
if you can come up with an .editorconfig that isn't too disruptive then I think that'd be fine

alexandercerutti commented:
I'll try to keep styles as similar as possible with the original code style 👍
I'll start a new branch starting from this one, so I'll format everything

piranna commented:
Should I start a branch too for the Int64 support of https://github.com/rvagg/bl/pull/109, or should I wait to get this merged first?

rvagg commented on package.json:
```diff
@@ -6,8 +6,9 @@
-    "test:ci": "npm run lint && node test/test.js",
+    "test": "npm run lint && npm run test:types && node test/test.js | faucet",
+    "test:ci": "npm run lint && node test/test.js && npm run test:types",
+    "test:types": "tsc --allowJs --noEmit test/test.js"
```
```suggestion
    "test:types": "tsc --allowJs --noEmit test/test.js",
```

rvagg commented on test/test.js:
```diff
@@ -501,7 +540,7 @@ tape('uninitialized memory', function (t) {
   const random = crypto.randomBytes(65534)
 
-  const bl = new BufferList((err, buf) => {
+    const bl = new BufferListStream((err, buf) => {
```
```suggestion
    const bl = new BufferListStream((err, buf) => {
```
